File: sorting/py_sort/heap.py
from __future__ import division


class MaxHeap:
    _heap = []

    def __init__(self, arr):
        for idx, entry in enumerate(arr):
            self._heap.append(entry)
            self._max_heapify_bottom_to_top(idx)

    def _max_heapify_bottom_to_top(self, idx):
        if idx == 0:
            return
        root = (idx - 1) // 2
        if self._heap[root] < self._heap[idx]:
            self._heap[root], self._heap[idx] = self._heap[idx], self._heap[root]
            self._max_heapify_bottom_to_top(root)

    def _max_heapify_top_to_bottom(self, idx):
        left = 2 * idx + 1
        right = 2 * idx + 2
        largest = idx
        arr = self._heap

        if left < len(arr) and arr[left] > arr[largest]:
            largest = left
        if right < len(arr) and arr[right] > arr[largest]:
            largest = right

        if largest != idx:
            self._heap[largest], self._heap[idx] = self._heap[idx], self._heap[largest]
            self._max_heapify_top_to_bottom(largest)

    def pop_root(self):
        if len(self.heap) == 1:
            return self._heap.pop(0)
        # swap root with last element, pop last element, normalize heap
        # less efficient version
        # self._heap[0], self._heap[self.last_idx] = self._heap[self.last_idx], self._heap[0]
        # prev_root = self._heap.pop()

        # more efficient version, only one assignment
        prev_root = self._heap.pop(0)
        self._heap.insert(0, self._heap.pop(self.last_idx))
        self._max_heapify_top_to_bottom(0)
        return prev_root

    def sorted(self):
        result = []
        while self.not_empty:
            result.append(self.pop_root())
        return result[::-1]

    @property
    def heap(self):
        return self._heap

    @property
    def last_idx(self):
        return len(self._heap) - 1

    @property
    def not_empty(self):
        return len(self._heap)


def heap_sort(data):
    return MaxHeap(data).sorted()

File: sorting/c_plus/heap.cpp
#include "heap.h"

class MaxHeap{
        void max_heapify_bottom_to_top(int idx);
        void max_heapify_top_to_bottom(int idx);
        int pop_root();
    public:
        vector <int> heap;
        MaxHeap (vector<int>& arr);
        vector<int> sorted();
};

MaxHeap::MaxHeap (vector<int>& arr){
    for (int i = 0; i < arr.size(); i++){
        heap.push_back(arr[i]);
        max_heapify_bottom_to_top(i);
    }
}

void MaxHeap::max_heapify_bottom_to_top(int idx){
    if (idx != 0)
    {
        int root_idx = (idx - 1) / 2;
        if (heap[root_idx] < heap[idx])
        {
            swap(heap[root_idx], heap[idx]);
            max_heapify_bottom_to_top(root_idx);
        }        
    }
}

void MaxHeap::max_heapify_top_to_bottom(int idx){
    int left_idx = 2 * idx + 1;
    int right_idx = 2 * idx + 2;
    int largest = idx;
    if (left_idx < heap.size() and heap[left_idx] > heap[largest])
    {
        largest = left_idx;
    }
    if (right_idx < heap.size() and heap[right_idx] > heap[largest])
    {
        largest = right_idx;
    }
    if (largest != idx)
    {
        swap(heap[largest], heap[idx]);
        max_heapify_top_to_bottom(largest);
    }
}

int MaxHeap::pop_root(){
    int prev_root;
    if (heap.size() == 1)
    {
        prev_root = heap.back();
        heap.pop_back();
    }
    else
    {
        swap(heap[0], heap[heap.size() -1]);
        prev_root = heap.back();
        heap.pop_back();
        max_heapify_top_to_bottom(0);
    }
    return prev_root;
}

vector<int> MaxHeap::sorted(){
    vector<int> result;
    while (!heap.empty())
    {
        result.push_back(pop_root());
    }
    reverse(result.begin(), result.end());
    return result;
}


void heap_sort(vector<int>& arr)
{
    MaxHeap max_heap (arr);
    arr = max_heap.sorted();
}
